### PR TITLE
[FLINK-33541][table-planner] function RAND and RAND_INTEGER should return type nullable if the arguments are nullable

### PR DESCRIPTION
**Pull request by xuyangzhong**
## What is the purpose of the change

Currently RAND and RAND_INTEGER are always return type "NOT NULL" as built in functions in `FlinkSqlOperatorTable`. But when codegen, the result type is always nullable in `RandCallGen`.

Actually, the result type of RAND and RAND_INTEGER should depend on the types of arguments. For example, If the argument is nullable, the return type should be nullable.

This bug causes the error with pattern "IF(1 = 1, RAND(0), 0)", because the following logic:
1. the result type of RAND is always "not null" when validation in `FlinkSqlOperatorTable`, and the third param is 0 that is always "not null", so the return type of operator IF is always "not null".
2. when codegen operator RAND, the return type of RAND(0) is always "nullable"
3. when codegen operator IF, the CAST rule cannot cast from "nullable" (the second arg RAND) to "not null"(the result type), so the exception throws.

## Brief change log

  - *correct the return type to be nullable when one of its args is nullable during validation*
  - *correct the final return type to be nullable when the original result type is nullable during codegen*
  - *add tests*

## Verifying this change

Some tests are added to verify this pr.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/FlinkSqlOperatorTable.java [hunk 1]
```diff
@@ -940,7 +940,9 @@ public SqlSyntax getSyntax() {
             new SqlFunction(
                     "RAND",
                     SqlKind.OTHER_FUNCTION,
-                    ReturnTypes.DOUBLE,
+                    ReturnTypes.cascade(
```
Comment (libenchao) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/FlinkSqlOperatorTable.java:943]: Shouldn't definitions in `BuiltInFunctionDefinitions` also be updated?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/RandFunctionTest.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+
+import java.util.stream.Stream;
+
+/**
+ * Test for {@link org.apache.flink.table.functions.BuiltInFunctionDefinitions#RAND} and {@link
+ * org.apache.flink.table.functions.BuiltInFunctionDefinitions#RAND_INTEGER} and their return type.
+ */
+public class RandFunctionTest extends BuiltInFunctionTestBase {
```
Comment (libenchao) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/RandFunctionTest.java:30]: all the tests in this package are `**ITCase`.

**Comment by xuyangzhong** [comment]
Sorry for the late apply, @schnappi17 . Can you do the final check about it? @libenchao

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala [hunk 3]
```diff
@@ -2661,6 +2665,37 @@ class ScalarFunctionsTest extends ScalarTypesTestBase {
     //      "1996-11-10 06:55:44.333")
   }
 
+  @Test
+  def testRandAndIf(): Unit = {
```
Comment (libenchao) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala:2669]: Currently we only have tests for rand with seed, actually the case reported in FLINK-33541 is without seed. I know currently the test suite cannot do this (check a expression/query runs without checking the expected result, or checking the result is in a range), so I won't have a strong opinion that we must add such a test case.

**Review by libenchao — APPROVED** [review]
@xuyangzhong Thanks for the updating, it looks good to me now. I've a comment about the test, but I'll leave the decision to you whether to address it.

**Comment by libenchao** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala (hunk 3 above)
Comment (xuyangzhong) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala:2669]: Let me take a look and will push again.

**Comment by xuyangzhong** [comment]
Hi, @libenchao sorry for this noise. I have updated this pr to test non-deterministic results about RAND and RAND_INTEGER. Could you have a look again?

**Review by libenchao — APPROVED** [review]
+1, merging

**Pull request closed without merging**